File: eyedb-bench/trunk/src/java/org/eyedb/benchmark/polepos/teams/eyedb/data/TreeUtil.java
package org.eyedb.benchmark.polepos.teams.eyedb.data;

public class TreeUtil {
    public static Tree createTree( Database database, int depth) throws org.eyedb.Exception
    {
	return createTree( database, depth, 0);
    }

    private static Tree createTree( Database database, int maxDepth, int currentDepth) throws org.eyedb.Exception
    {
	if (currentDepth >= maxDepth)
	    return null;

	Tree tree = new Tree( database);

	if (currentDepth == 0)
	    tree.setName( "root");
	else
	    tree.setName( "node at depth " + currentDepth);

	tree.setDepth( currentDepth);

	tree.setPreceding( createTree( database, maxDepth, currentDepth + 1));
	tree.setSubsequent( createTree( database, maxDepth, currentDepth + 1));

	return tree;
    }

    public static void traverse( Tree tree, TreeVisitor visitor) throws org.eyedb.Exception
    {
        if (tree == null)
            return;

        traverse( tree.getPreceding(), visitor);
        traverse( tree.getSubsequent(), visitor);
        visitor.visit(tree);
    }
}

File: eyedb-bench/trunk/src/cpp/polepos/eyedb/treeutil.cc
#include <sstream>
#include "treeutil.h"

#include <iostream>

using namespace std;

Tree * TreeUtil::createTree( eyedb::Database *database, int depth) throw( eyedb::Exception)
{
  return createTree( database, depth, 0);
}

void TreeUtil::traverse( Tree *tree, TreeVisitor &visitor) throw( eyedb::Exception)
{
  if (tree == 0)
    return;

  traverse( tree->getPreceding(), visitor);
  traverse( tree->getSubsequent(), visitor);
  visitor.visit(tree);
}

Tree * TreeUtil::createTree( eyedb::Database *database, int maxDepth, int currentDepth) throw( eyedb::Exception)
{
  if (currentDepth >= maxDepth)
    return 0;
      
  Tree *tree = new Tree( database);

  if (currentDepth == 0)
    tree->setName( "root");
  else {
    ostringstream oss;

    oss << "node at depth " <<  currentDepth;
    
    tree->setName( oss.str());
  }

  tree->setDepth( currentDepth);

  tree->setPreceding( createTree( database, maxDepth, currentDepth + 1));
  tree->setSubsequent( createTree( database, maxDepth, currentDepth + 1));

  return tree;
}

class PrintTreeVisitor : public TreeVisitor {
public:
  void visit( Tree * tree);
};

void PrintTreeVisitor::visit( Tree * tree)
{
  try {
    cout << "Tree: " << tree->getDepth() << " \"" << tree->getName() << "\"" << endl;
  } catch (eyedb::Exception &ex) {
    ex.print();
  }
}

void TreeUtil::print( Tree *tree)
{
    PrintTreeVisitor visitor;
    TreeUtil::traverse( tree, visitor);
}

